### ISSUE TITLE
slice error

### ISSUE THREAD

**Issue by chenbohua3** [pytorch/pytorch#20149]
## 🐛 Bug

different behavior of slicing tensor with slices in different tensor type
## To Reproduce

```
import torch

a = torch.randint(low=1, high=10, size=(2, 3))
print(a)
b = torch.randint(2, (3,), dtype=torch.uint8)
print(b)
a[0, b] = 0
print(a)

```

## Expected behavior

if dtype of b is torch.uint8, it works normally like:
```
tensor([[9, 5, 3],
        [1, 3, 9]])
tensor([1, 0, 1], dtype=torch.uint8)
tensor([[0, 5, 0],
        [1, 3, 9]])
```
if I change dtype to torch.int64, it woks differently like:
```
tensor([[3, 9, 9],
        [8, 7, 7]])
tensor([1, 0, 1])
tensor([[0, 0, 9],
        [8, 7, 7]])
```
it seems that the value of slices act as slices instead of the position of no-zero elements
## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Mac os 10.14.3
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: None
 - GPU models and configuration:  None
 - Any other relevant information:  None

**Comment by izdeby** [comment]
@chenbohua3, can you, please, verify that you are still getting this behavior? I can't repro the issue: 
![Screen Shot 2019-05-06 at 11 50 12 AM](https://user-images.githubusercontent.com/47012416/57237416-2af6d180-6ff5-11e9-9f72-0a82b98d1cb9.png)

**Comment by fmassa** [comment]
I believe this behavior is expected: until we disable boolean indexing with `uint8` tensors, this behavior will persist. This was btw one of the reasons why we introduced `bool` tensors in the first place, so that this becomes consistent.

This means that, for now, indexing with a `torch.tensor([0, 1, 1], dtype=torch.uint8)` tensor and with a `torch.tensor([0, 10, 20], dtype=torch.uint8)` gives the same results, and the `torch.tensor([0, 10, 20], dtype=torch.uint8)` is considered as `[False, True, True]`.

**Comment by ezyang** [comment]
Just to clarify what @fmassa said, the uint8 behavior is *not* the expected behavior; if you pass in a tensor to the index, we're supposed to interpret the tensor values as indices you want to index out, and not as a boolean mask. The fact that uint8 is interpreted as a boolean mask is something we expect to change eventually. We can leave this issue open for now to keep track of the matter.

cc @izdeby

**Comment by chenbohua3** [comment]
@izdeby yes, I can reproduce this problem. It seems that `int64 b` in your codes happens to be [1, 1, 0], which will lead to consistent results of boolean mask and index. Try to make `int64 b` to [1, 0, 1], this problem will be reproduced.

**Comment by lsb** [comment]
Currently seeing this behavior still. Also I don't understand the error message I get when I use a `int8` or other type:

```
torch.tensor([1, 2, 3], dtype=torch.int64)[torch.tensor([0,1,2,1], dtype=torch.int8)]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[8], [line 1]
----> [1] torch.tensor([1, 2, 3], dtype=torch.int64)[torch.tensor([0,1,2,1], dtype=torch.int8)]

IndexError: tensors used as indices must be long, int, byte or bool tensors
```

I didn't see how best to index into an 8-bit lookup table without converting the indices into int32s.

Furthermore, torch.gather requires indices to be int64, which is further double the memory usage/bandwidth.

Thanks for all your efforts in making PyTorch! Much appreciated
